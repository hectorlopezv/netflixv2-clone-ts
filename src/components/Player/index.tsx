import React, {useState, useContext, createContext} from 'react';
//here is the big MODAL using React Portal
import ReactDom from 'react-dom';
import {Container, Button, Overlay, Inner, Close} from './styles/Player';

export interface PlayerProps {
    
}
const PlayerContext = createContext<{[prop: string]: any}>({});
const Player: React.FC<PlayerProps> = ({children, ...props}) => {
    const [showPlayer, setshowPlayer] = useState(false);

    return (  
        <PlayerContext.Provider value={{showPlayer, setshowPlayer}}>
            <Container {...props}>{children}</Container>
        </PlayerContext.Provider> 
    );
}

export default Player;


export interface Player_VideoProps {
    src: string;
}
 
export const Player_Video: React.FC<Player_VideoProps> = ({src, ...props}) => {
    const { showPlayer, setshowPlayer } = useContext(PlayerContext);
    return showPlayer ? ReactDom.createPortal(
    <Overlay data-testid="player" onClick={() => setshowPlayer(false)  }>
        <Inner>
            <video id="netflix-player" controls>
                <source src={src} type="video/mp4" />
            </video>
            <Close />
        </Inner>
    </Overlay>, document.body): null;
}


export interface Player_ButtonProps {
    
}
 
export const Player_Button: React.FC<Player_ButtonProps> = () => {
    const { showPlayer, setshowPlayer } = useContext(PlayerContext);

    return <Button onClick={() => setshowPlayer( (prevstate: boolean) => !prevstate )}>
        Play
    </Button>
}
 

 