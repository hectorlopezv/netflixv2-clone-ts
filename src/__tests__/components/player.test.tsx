/* eslint-disable react/jsx-pascal-case */
import React from 'react'

import { render, fireEvent } from '@testing-library/react';
import Player , { Player_Button, Player_Video} from '../../components/Player';

///we ca use Enzyme
//Jest Magic
//Describre --> beforeEach() --> it(  --> expect())

describe('Testing Player Componnet <Player />', () => {

    it('renders the <Player /> with a bunny video', () => {
        //we are testing components
        //we prefer end to end test (all funcitonality)
        const {container, getByText, queryByTestId, debug} = render(
            <Player>
                <Player_Button />
                <Player_Video src="videos/bunny.mp4"/>
            </Player>
        );
        //debugg
        //debug();
        //to set data-testid="player" for the componnetn to be teste
        expect(queryByTestId('player')).toBeFalsy();
        //fire event play click on play component PLAY DE VIDEO
        fireEvent.click(getByText('Play'));

        expect(queryByTestId('player')).toBeTruthy();
        //pause video
        fireEvent.click(queryByTestId('player')!);

        expect(queryByTestId('player')).toBeFalsy();

        expect(container.firstChild).toMatchSnapshot();
    });

});
