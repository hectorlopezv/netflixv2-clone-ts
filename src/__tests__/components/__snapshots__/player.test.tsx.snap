// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Player Componnet <Player /> renders the <Player /> with a bunny video 1`] = `
<div
  class="Player__Container-f42nev-0"
>
  <button
    class="Player__Button-f42nev-1 fblUjx"
  >
    Play
  </button>
</div>
`;
